File: ios/ReservationApp/ReservationApp/Presentation/ViewModels/LoginViewModel.swift
//
//  LoginViewModel.swift
//  ReservationApp
//
//  Created by Mehmet Ali Kısacık on 13.11.2025.
//

import Foundation
import SwiftUI
import Combine
enum LoginRole: String, CaseIterable {
    case personel = "Personel"
    case admin = "Admin"
}

@MainActor
class LoginViewModel: ObservableObject {
    @Published var email: String = ""
    @Published var password: String = ""
    @Published var selectedRole: LoginRole = .personel
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var isAuthenticated: Bool = false
    @Published var currentUser: User?
    
    private let authRepository: AuthRepositoryProtocol
    private let keychainManager: KeychainManager
    
    var onLoginSuccess: ((User) -> Void)?
    
    init(
        authRepository: AuthRepositoryProtocol? = nil,
        keychainManager: KeychainManager? = nil
    ) {
        // Create dependencies on main actor to avoid concurrency issues
        self.authRepository = authRepository ?? AuthRepository()
        self.keychainManager = keychainManager ?? KeychainManager.shared
    }
    
    func login() async {
        // Clear previous error
        errorMessage = nil
        
        // Basic validation
        guard !email.isEmpty, !password.isEmpty else {
            errorMessage = "Lütfen email ve şifre girin"
            return
        }
        
        // Email format validation
        guard isValidEmail(email) else {
            errorMessage = "Geçerli bir email adresi girin"
            return
        }
        
        isLoading = true
        
        do {
            // Login request - token is saved by repository
            _ = try await authRepository.login(email: email, password: password)
            
            // Token should be saved by repository, but verify
            guard keychainManager.hasToken() else {
                errorMessage = "Giriş başarısız. Token alınamadı."
                isLoading = false
                return
            }
            
            // Fetch current user
            let user = try await authRepository.getCurrentUser()
            
            // Validate role matches selected role
            let userRole = user.role
            let isUserAdmin = userRole == .admin
            let selectedIsAdmin = selectedRole == .admin
            
            if isUserAdmin && !selectedIsAdmin {
                keychainManager.deleteToken()
                errorMessage = "Bu hesap Admin hesabıdır. Admin bölümünden giriş yapmalısınız."
                isLoading = false
                return
            } else if !isUserAdmin && selectedIsAdmin {
                keychainManager.deleteToken()
                errorMessage = "Bu hesap Personel hesabıdır. Personel bölümünden giriş yapmalısınız."
                isLoading = false
                return
            }
            
            // Success - set user and notify
            currentUser = user
            isAuthenticated = true
            onLoginSuccess?(user)
            
        } catch let error as NetworkError {
            errorMessage = error.errorDescription ?? "Giriş başarısız. Email ve şifrenizi kontrol edin."
        } catch {
            errorMessage = "Bir hata oluştu. Lütfen tekrar deneyin."
        }
        
        isLoading = false
    }
    
    private func isValidEmail(_ email: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
}

File: android/feature/auth/src/main/java/com/reservationapp/feature/auth/viewmodel/LoginViewModel.kt
package com.reservationapp.feature.auth.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.reservationapp.core.common.AuthStateManager
import com.reservationapp.core.common.Result
import com.reservationapp.core.common.extensions.isValidEmail
import com.reservationapp.domain.model.User
import com.reservationapp.domain.repository.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class LoginRole {
    PERSONEL,
    ADMIN
}

class LoginViewModel(
    private val authRepository: AuthRepository,
    private val authStateManager: AuthStateManager
) : ViewModel() {

    val email = MutableStateFlow("")
    val password = MutableStateFlow("")
    val selectedRole = MutableStateFlow(LoginRole.PERSONEL)

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    fun login() {
        _errorMessage.value = null

        if (email.value.isEmpty() || password.value.isEmpty()) {
            _errorMessage.value = "Lütfen email ve şifre girin"
            return
        }

        if (!email.value.isValidEmail()) {
            _errorMessage.value = "Geçerli bir email adresi girin"
            return
        }

        _isLoading.value = true

        viewModelScope.launch {
            when (val result = authRepository.login(email.value, password.value)) {
                is Result.Success -> {
                    val authResponse = result.data

                    if (authResponse.authToken.isEmpty()) {
                        _errorMessage.value = "Giriş başarısız. Token alınamadı."
                        _isLoading.value = false
                        return@launch
                    }

                    when (val userResult = authRepository.getCurrentUser()) {
                        is Result.Success -> {
                            val user = userResult.data

                            val isUserAdmin = user.role == "Admin"
                            val selectedIsAdmin = selectedRole.value == LoginRole.ADMIN

                            if (isUserAdmin && !selectedIsAdmin) {
                                authRepository.logout()
                                _errorMessage.value = "Bu hesap Admin hesabıdır. Admin bölümünden giriş yapmalısınız."
                                _isLoading.value = false
                                return@launch
                            } else if (!isUserAdmin && selectedIsAdmin) {
                                authRepository.logout()
                                _errorMessage.value = "Bu hesap Personel hesabıdır. Personel bölümünden giriş yapmalısınız."
                                _isLoading.value = false
                                return@launch
                            }

                            _currentUser.value = user
                            _isAuthenticated.value = true
                            authStateManager.setCurrentUser(user)
                            authStateManager.setAuthenticated(true)
                        }
                        is Result.Error -> {
                            _errorMessage.value = "Kullanıcı bilgileri alınamadı."
                        }
                        is Result.Loading -> {}
                    }
                }
                is Result.Error -> {
                    _errorMessage.value = when (result.exception) {
                        is com.reservationapp.core.network.NetworkError.Unauthorized -> {
                            "Giriş başarısız. Email ve şifrenizi kontrol edin."
                        }
                        else -> {
                            "Bir hata oluştu. Lütfen tekrar deneyin."
                        }
                    }
                }
                is Result.Loading -> {}
            }

            _isLoading.value = false
        }
    }
}

